[WIP] create forms in the requested context

antis81 opened the pull request:
This PR improves general resource management within the UI. While not the main purpose, It also has a positive effect on qTox's memory footprint. The goal is to mainly improve resource management, where camera & audio resources are requested/free'd.

So, instead of creating all widgets at program start, they are now created within the context and free'd, when "closed". That means, a widget is not in memory until the user requests it. To make it complete for settings-ui however, we need to introduce a widget factory, that creates the widget for a specific "settings tab".

TODO:
- [ ] Destroy widgets in "embedded window" mode, when they are replaced. (**big change**, but totally worth it) 
- [ ] Implement a widget factory for "settings tab" widgets, so that only the "visible" widget is in memory.

@Diadlo, @initramfs, @sudden6 FYI

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tux3/qtox/3564)

<!-- Reviewable:end -->

antis81 committed:
refactor(widget): fix static analyzer warning and some cleanup

antis81 committed:
refactor(widget): wrap content layout in a "weak" pointer

Like every layout, the content layout is parented and owned by a widget. This owner is responsible for layout destruction.

zetok commented:
on 0d8d46457c9ef6fb270ae1cb97f7fb2088740cd7
- [ ] lot of spam in stdout that looks like that
  
  ```
  [15:58:38.417 UTC] kernel/qobject.cpp:4657 : Warning: QObject::connect: invalid null parameter
  [15:58:38.472 UTC] kernel/qobject.cpp:4657 : Warning: QObject::connect: invalid null parameter
  [15:58:38.504 UTC] kernel/qobject.cpp:4657 : Warning: QObject::connect: invalid null parameter
  [15:58:38.536 UTC] kernel/qobject.cpp:4657 : Warning: QObject::connect: invalid null parameter
  ```
- [ ] there's no top settings icon when first clicked on settings for the first time, switching tabs brings them back
  ![no top bar](https://cloud.githubusercontent.com/assets/3148759/17271581/8fb18ccc-5677-11e6-9a92-8a79187ef6a9.png)
  ![with bar](https://cloud.githubusercontent.com/assets/3148759/17271592/cb88c968-5677-11e6-9956-d19f8fd12267.png)

I didn't really test whether changing settings did not regress.

antis81 commented:
Thanks for running  this shortly.

> lot of spam in stdout that looks like that

Thanks -> solved! The magic happens in `Settings`, which now correctly emits the signals when the value changes. This solves quite some "connection knots" by centralizing the settings values. That means, instead of connecting slots to gui component (e.g. a QCheckBox), we connect to the settings value "directly". If that changes (and only then!), the value updates and notifies safely. This also works fine to connect independent widgets on the same settings value! Anyway, I did not consequently implement it for all the values (just for the ones affected by this PR), because it might interfere with Diadlo's PR (read below), that cleans up the settings.

> there's no top settings icon when first clicked on settings for the first time, switching tabs brings them back

I reverted a change and will solve that. Won't that title bar vanish soon anyway (see #3457)?

sudden6 commented:
@antis81 I think you referenced a wrong issue

antis81 commented:
@sudden6 Argh damn, once again, thanks. Will correct that of course.

antis81 committed:
feat(settings): create ui when needed instead of show/hide

note: no change in "embedded window" mode yet, but doesn't harm either

initramfs commented on src/core/recursivesignalblocker.cpp:
```diff
@@ -0,0 +1,64 @@
+/*
+    Copyright © 2016-2017 by The qTox Project
```
According to FSF & GPL, you can't "pre-license" a year in which the software hasn't been released for (2017). 

> The copyright notice should include the year in which you finished preparing the release (so if you finished it in 1998 but didn't post it until 1999, use 1998). You should add the proper year for each release; for example, “Copyright 1998, 1999 Terry Jones” if some versions were finished in 1998 and some were finished in 1999.

So I suggest you change the line to just:

```
Copyright © 2016 by The qTox Project
```

[Source link](https://www.gnu.org/licenses/gpl-howto.html)

initramfs commented on src/core/recursivesignalblocker.h:
```diff
@@ -0,0 +1,44 @@
+/*
+    Copyright © 2016-2017 by The qTox Project
```
Same copyright issue as the other file.

zetok commented:
@antis81 

> Won't that title bar vanish soon anyway (see #3457)?

Yes, but this here is a bug, and provides only inconsistent behavior.

antis81 committed:
refactor: no need to hide widgets in layout clear

antis81 commented:
@zetok Initially, I just wanted to create/destroy settings widget, but it totally does not work, because this crap `ContentLayout` comes in the way. It doesn't make sense to run it right now. Please stay tuned while I'm rewriting the layout stuff (~90% complete). It will also fix layout, when the friends list is near the right or left edge of the screen.

antis81 committed:
refactor: cleanup

* enum class FilterCriteria
* prepare widgets with guarded pointers
* alphabetic order in class forward declaration

antis81 committed:
feat(settings): add RecursiveSignalBlocker

can be used to block any signals during settings ui initialization

antis81 committed:
refactor(gui): add ContentWidget and some helper methods

* add "Widget::minimumSizeHint" overload
* add generic method "Widget::showContentForm":
  The method takes care of window creation and placement within desktop/screen borders. It will be used in upcoming commits.

TODO:
* While working, widget placement is not implemented yet for "embedded mode" (splitter).

antis81 committed:
fix(gui): connect change notifications to settings instead of settings widget

Also cleans up some code around that, which doesn't make sense to put into separate commit.
